Simplify counting in AbstractSearchAsyncAction

original-brownbear opened the pull request:
No need to do this so complicated, just count down one when we're actually done with a specific shard id.

original-brownbear committed:
Simplify counting in AbstractSearchAsyncAction

No need to do this so complicated, just count down one when we're actually done with a specific shard id.

elasticsearchmachine commented:
Pinging @elastic/es-search-foundations (Team:Search Foundations)

original-brownbear commented:
For others: This is a nice step on the way to #118490 (in addition to just being a considerable simplification) since it aligns the counting with the way it's done for batched execution in that PR.

original-brownbear commented:
This also removes all uses of the group shards iterator interfaces and makes https://github.com/elastic/elasticsearch/pull/116891 a trivial to understand cleanup.

original-brownbear commented on server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java:
```diff
@@ -449,17 +454,11 @@ private void onShardFailure(final int shardIndex, SearchShardTarget shard, final
-        if (totalOps == expectedTotalOps) {
-            onPhaseDone();
-        } else if (totalOps > expectedTotalOps) {
-            throw new AssertionError(
```
No need for this. It's unclear if you'd want to kill a node if there actually was a bug here and for tests this is asserted down the line by checking listener double resolves.

original-brownbear committed:
Merge branch 'main' into simplify-abstract-async-search

original-brownbear commented on server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java:
```diff
@@ -90,15 +92,25 @@ abstract class AbstractSearchAsyncAction<Result extends SearchPhaseResult> exten
+
+    static {
+        try {
+            OUTSTANDING_SHARDS = MethodHandles.lookup().findVarHandle(AbstractSearchAsyncAction.class, "outstandingShards", int.class);
```
The constructor here is confusingly expensive in profiling, maybe starting to relax the ordering a little will help so I figured I'd start here while I'm changing this anyway and at least go to just release style semantics in the constructor.

javanna replied:
I understand that skipped shards do not need their own counter as they are known in advance. I also agree that moving to counting each shard as 1 is simpler, and then there is no need to compare expected ops with total ops. 

Can we do without the `VarHandle` and use a simple `AtomicInteger` that we decrement each time though? What makes the var handle needed in this scenario?

original-brownbear replied:
Right, the var handle came out of a version of this change that went a little further lets say :) with the way it works now, the atomic integer might also be faster, no need to put the changing counter in a cacheline with other read-only fields :) => atomic int it is

javanna commented on server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java:
```diff
@@ -139,18 +151,12 @@ abstract class AbstractSearchAsyncAction<Result extends SearchPhaseResult> exten
-        // it's number of active shards but use 1 as the default if no replica of a shard is active at this point.
-        // on a per shards level we use shardIt.remaining() to increment the totalOps pointer but add 1 for the current shard result
-        // we process hence we add one for the non active partition here.
-        this.expectedTotalOps = shardsIts.totalSizeWith1ForEmpty();
```
For posterity, this was introduced with https://github.com/elastic/elasticsearch/commit/faefc772a462546b0127b674c8b6d3dccfd61c7e . The intent was to account for inactive shards (that are in the process of being allocated) as 1, so that they are expected to fail, otherwise their failure would make the search complete earlier than expected, missing results from other active shards.

Together with it `SearchWhileCreatingIndexTests` was added, which is now called `SearchWhileCreatingIndexIT`. The test went through a lot of changes and refactoring over time, and it was also quite problematic and flaky in the early days. Funnily enough, if I remove the counting of the empty group as 1 (and call `totalSize` instead), this specific test still succeeds. I may not have run it enough times to cause failures, or perhaps the issue that this was fixing no longer manifests. Either way, a lot of other tests fail due to too many ops executed compared to the expected ops, because the counting also needs to be adjusted accordingly (which is expected).

In principle, I agree that counting each shard as 1, regardless of how many copies it has whether that be inactive, primary only, one replica or multiple replicas is simpler.

javanna commented on server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java:
```diff
@@ -449,17 +454,11 @@ private void onShardFailure(final int shardIndex, SearchShardTarget shard, final
-            if (lastShard == false) {
-                performPhaseOnShard(shardIndex, shardIt, nextShard);
+            if ((int) OUTSTANDING_SHARDS.getAndAdd(this, -1) == 1) {
+                onPhaseDone();
```
With the updated counting, a failure is counted only if all copies of the same shard fails, correct? Otherwise, one shard is counted when its operation succeeds. That is the reason why you flipped the conditional here? Maybe worth adding a small comment about it.

javanna commented on server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java:
```diff
@@ -552,28 +551,12 @@ private void onShardResultConsumed(Result result, SearchShardIterator shardIt) {
-            // but its range was available in the IndexMetadata, in that
-            // case the shardsIt.remaining() would be 0, expectedTotalOps
-            // accounts for unavailable shards too.
-            remainingOpsOnIterator = Math.max(shardsIt.remaining(), 1);
```
definitely good to get rid of this special case.

javanna commented on server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java:
```diff
@@ -552,28 +551,12 @@ private void onShardResultConsumed(Result result, SearchShardIterator shardIt) {
-            throw new AssertionError(
-                "unexpected higher total ops [" + xTotalOps + "] compared to expected [" + expectedTotalOps + "]",
-                new SearchPhaseExecutionException(getName(), "Shard failures", null, buildShardFailures())
-            );
```
Here too, I think I would restore this as an assert, just to make sure.

javanna reviewed (commented):
I left a few comments. Shall we remove the two unused methods from `GroupShardsIterator` as well (`totalSize` and `totalSizeWith1ForEmpty`) ?

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into simplify-abstract-async-search

original-brownbear committed:
comment and assert

original-brownbear committed:
Merge remote-tracking branch 'origin/simplify-abstract-async-search' into simplify-abstract-async-search

original-brownbear commented:
@javanna can we just remove the whole iterator class next as suggested in #116891 ? :) It's an absolutely trivial change after this, the methods you mention are essentially all there is to that class today?

original-brownbear committed:
atomic

javanna commented on server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java:
```diff
@@ -449,17 +440,14 @@ private void onShardFailure(final int shardIndex, SearchShardTarget shard, final
-            if (lastShard == false) {
-                performPhaseOnShard(shardIndex, shardIt, nextShard);
+            // count down outstanding shards, we're done with this shard as there's no more copies to try
+            final int outstanding = outstandingShards.getAndDecrement();
```
minor nit: is there a specific reason that you do getAndDecrement as opposed to decrementAndGet here? I think that for readability `if (outstanding == 0)` would be directly clearer here, together with asserting that outstanding can't go negative.

javanna reviewed (approved):
left a minor comment for readability. I would still remove the unused methods from GroupShardsIterator, leaving the removal of the entire class for a follow-up. If you prefer to leave the methods and do the removal later, that's also good, but I would not do it in this PR.

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into simplify-abstract-async-search

original-brownbear committed:
cleanup

original-brownbear commented:
Thanks Luca :)

Removed the methods now and fixed the `decrementAndGet` (only used the other direction before because of the var handle, much cleaner now) I'll update my follow up PR for the removal of the class :)

original-brownbear commented:
@javanna I updated https://github.com/elastic/elasticsearch/pull/116891 which should be a trivial change after this one :)
